x/crosschain/keeper/grpc_query_cctx.go

```diff
@@ -161,19 +154,266 @@ func (k Keeper) CctxListPending(c context.Context, req *types.QueryListCctxPendi
 
 	// now query the pending nonces that we know are pending
 	for i := pendingNonces.NonceLow; i < pendingNonces.NonceHigh && !maxCCTXsReached(); i++ {
-		nonceToCctx, found := k.GetObserverKeeper().GetNonceToCctx(ctx, tss.TssPubkey, req.ChainId, i)
-		if !found {
-			return nil, status.Error(codes.Internal, "nonceToCctx not found")
-		}
-		cctx, found := k.GetCrossChainTx(ctx, nonceToCctx.CctxIndex)
-		if !found {
-			return nil, status.Error(codes.Internal, "cctxIndex not found")
+		cctx, err := getCctxByChainIDAndNonce(k, ctx, tss.TssPubkey, req.ChainId, i)
+		if err != nil {
+			return nil, err
 		}
-		cctxs = append(cctxs, &cctx)
+		cctxs = append(cctxs, cctx)
 	}
 
 	return &types.QueryListCctxPendingResponse{
 		CrossChainTx: cctxs,
 		TotalPending: totalPending,
 	}, nil
 }
+
+// CctxListPendingWithinRateLimit returns a list of pending cctxs that do not exceed the outbound rate limit
+// a limit for the number of cctxs to return can be specified or the default is MaxPendingCctxs
+func (k Keeper) CctxListPendingWithinRateLimit(c context.Context, req *types.QueryListCctxPendingWithRateLimitRequest) (*types.QueryListCctxPendingWithRateLimitResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	// check limit and use default MaxPendingCctxs if not specified
+	if req.Limit > MaxPendingCctxs {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("limit exceeds max limit of %d", MaxPendingCctxs))
+	}
+	limit := req.Limit
+	if limit == 0 {
+		limit = MaxPendingCctxs
+	}
+
+	// get current height and tss
+	ctx := sdk.UnwrapSDKContext(c)
+	height := ctx.BlockHeight()
+	if height <= 0 {
+		return nil, status.Error(codes.OutOfRange, "height out of range")
+	}
+	tss, found := k.zetaObserverKeeper.GetTSS(ctx)
+	if !found {
+		return nil, status.Error(codes.Internal, "tss not found")
+	}
+
+	// check rate limit flags to decide if we should apply rate limit
+	applyLimit := true
+	rateLimitFlags, found := k.GetRateLimiterFlags(ctx)
+	if !found || !rateLimitFlags.Enabled {
+		applyLimit = false
+	}
```

Comment:
It might be a good idea to add this under crosschain-flags

Reply:
I think separate flag probably works better if it's intended for a very different purpose.